### ISSUE TITLE
Update codebase's types to use records

### ISSUE THREAD

**Issue by JordanMartinez** [purescript/purescript#4115]
See [this comment in #4100](https://github.com/purescript/purescript/pull/4100#discussion_r652344792) for context.

I think this would best be done a few types at a time. Otherwise, we may get merge conflicts from other PRs that invalidates the work made towards this issue.

**Comment by hdgarrood** [comment]
Yeah, I think there are a few places where we have data types or data constructors with 3+ arguments where we’d really benefit from named record fields.

**Comment by JordanMartinez** [comment]
With #4351 merged, we can now update the codebase to use the [`RecordDotSyntax`](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0282-record-dot-syntax.rst). Since this issue was already about migrating more of our types to used records, coupling these two efforts into one issue seems reasonable to me.

**Comment by JordanMartinez** [comment]
`RecordDotSyntax` is actually two extensions:
- [`OverloadedRecordDot`](https://downloads.haskell.org/~ghc/9.2.3/docs/html/users_guide/exts/hasfield.html#solving-hasfield-constraints), which enables getters for records
- [`OverloadedRecordUpdate`](https://downloads.haskell.org/~ghc/9.2.3/docs/html/users_guide/exts/overloaded_record_update.html), which enables (nested) setters for records. However, enabling this extension currently requires us to also enable [`RebindableSyntax`](https://downloads.haskell.org/~ghc/9.2.3/docs/html/users_guide/exts/rebindable_syntax.html?highlight=rebindablesyntax#extension-RebindableSyntax)

Should we only enable `OverloadedRecordDot` for the time being until a version of GHC no longer requires enabling `OverloadedRecordUpdate` to also enable `RebindableSyntax`? Or is this not an issue?
